### PR TITLE
Implement `store_perms` & `store_path_perms` to specify file permissions

### PR DESCRIPTION
**Pull request by HorlogeSkynet**
Hello :wave:

Summary : This patch implements two new functions, allowing to pass file permissions to `store*` functions.

Rationale : Software configuration files usually contain secrets that should not be world-readable (default `umask` on most of operating systems).

As Rust does not support optional parameters, I've opted for fully-separate functions, while keeping _diff_ simple by renaming current `store_path` function to (internal) `do_store`.

A test case has been added with UNIX file permissions.

Feel free to edit this branch as needed.

Bye :pray:

### CODE REVIEW

File: src/lib.rs
Comment (Zykino) [inline src/lib.rs:314]: In rust the `pub` keyword is enough to know if a function is public or private. No need for an underscore prefix
```suggestion
fn store<T: Serialize>(
```

File: src/lib.rs
Comment (Zykino) [inline src/lib.rs:315]: Since the function is private, we can avoid compilation time/bloat by removing the generic argument. The public functions should transform the generic arguments into concrete ones.
```suggestion
    path: &Path,
```

File: src/lib.rs
Comment (Zykino) [inline src/lib.rs:349]: I guess this is not an `OpenConfigurationFileError` but a `SetPermissionsFileError`

**Review by Zykino — CHANGES_REQUESTED** [review]
Hello, thanks for the interest! I did some comments on the PR.

I am not sure this is good to include into `confy`: This change does not work on Windows. It also expect an understanding of Unix permissions.
Also users of this crates can trivially wrap this logic since they either give the configuration path or can retrieve it with `get_configuration_file_path`. This can permit them to think about how to handle permissions on each OS.

File: src/lib.rs
Comment (HorlogeSkynet) [reply src/lib.rs]: `store` was already taken, I've renamed it to `do_store` (lacking of better idea).

**Comment by HorlogeSkynet** [comment]
> Hello, thanks for the interest! I did some comments on the PR.

Thanks for your quick review !

> I am not sure this is good to include into `confy`: This change does not work on Windows.

`set_permissions` is not UNIX-specific. According to [documentation](https://doc.rust-lang.org/std/fs/fn.set_permissions.html) it also corresponds to a low-level function on Windows.

> It also expect an understanding of Unix permissions.

Yes, but previous (and simple) API will be still there if one doesn't know what they must/should set :slightly_smiling_face: 

> Also users of this crates can trivially wrap this logic since they either give the configuration path or can retrieve it with `get_configuration_file_path`. This can permit them to think about how to handle permissions on each OS.

Yes, but we _lose_ atomicity of saving configuration/secrets under the correct permissions (there is a time-window where data are written on disk with potentially too broad permissions, which is not the case when `fchmod` is directly used on file descriptor).

---

Bye :wave:

**Comment by deg4uss3r** [comment]
Hello and thank you for this PR, overall I do like this change as it is non-breaking and adds some functionality into this API that makes sense. 

My one follow up is if you could also write a Windows platform test that does the same as the unix one and checks the readonly metadata on a config file so they don't get out-of-sync.

**Comment by HorlogeSkynet** [comment]
Thanks for your message @deg4uss3r, I'll try to add such a test case soon™. Bye 👋

**Comment by HorlogeSkynet** [comment]
Hey @deg4uss3r, I've added a `test_store_path_perms_readonly` test case which deals with `std::fs::Permissions::readonly` (which is cross-platform). Tell me what you think about that ! :pray:

**Review by deg4uss3r — APPROVED** [review]
Approving the changes now with the additional tests added, the PR is not specific to UNIX (my initial concern as well) and does not break previous APIs the additional file permissions/metadata is an optional different path, nor does this change the file permissions on a newly created file using the old path it's all the same as before.

Thank you for this addition!

**Comment by deg4uss3r** [comment]
@Zykino Let me know if this addresses your original concerns if not happy to discuss more before merging.

**Review by Zykino — APPROVED** [review]
Sorry I wanted to help with an early review and wasn’t aware that on github that would require me to follow through. I have a little nit… but won’t block on that.

I’m a bit surprised by the fact that you can set a file to readonly and then write to it. But apparently rust (or the OS?) authorize it, since this property is checked at the moment we open the file and not at the call to write. See this [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=adb6f46545f0d4d8f43f91d70c12632b) if I’m not clear.

File: src/lib.rs
Comment (HorlogeSkynet) [reply src/lib.rs:465]: Unfortunately there is a nit about this nit : test will fail on (UNIX) hosts with an umask of `0o222` (or "above"), as new files will be read-only by default.

Comment (Zykino) [reply src/lib.rs:465]: Even between 2 stores ? I’ll have to believe you there.

I’m not sure what you test with 2 stores one after the other like that. The first store (a line above)
https://github.com/rust-cli/confy/blob/9b6da7929e98c5676a310a4900f8c0d0f8b978ae/src/lib.rs#L464
should open, write the file and close it.

So at that point the file is closed and you want to write to it again, but setting permissions.

Maybe you did not want to store to file twice? (That is also a good "fix" to only test 1 store : the perm one.)

Anyway to me the PR is good, I learned some too :)

Comment (HorlogeSkynet) [reply src/lib.rs:465]: > Even between 2 stores ? I’ll have to believe you there.

On an UNIX host :

1. Storing without specifying permissions with an umask of `0o222` causes files to miss writing bits --> `readonly` attribute will evaluate to `false`, hence adding such a test would fail.
2. Storing and specifying `readonly` will always succeed (no umask could prevent this) --> proposed test is environment-safe.

However, I don't have your project testing policy in mind, and maybe you don't expect unit tests to pass in such environments.

> Maybe you did not want to store to file twice? (That is also a good "fix" to only test 1 store : the perm one.)

I did want to, but Rust standard library does not allow `std::fs::Permissions` direct instantiation, so I've opted for a snippet where we could safely switch the `readonly` attribute from metadata read from a file descriptor we know we could open in this testing environment.
I could have modified `test_store_path` case but didn't want to update existing tests.

Thanks for your time, bye :wave:

**Comment by deg4uss3r** [comment]
Merged, thank you!

**Pull request merged**
